File: output/python/tests/31_doc_lamp_basic.py

# Documentation Example: Basic Lamp with enter/exit events


from typing import Any, Optional, List, Dict, Callable

class LampFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters


class LampFrameContext:
    def __init__(self, event: LampFrameEvent, default_return):
        self.event = event
        self._return = default_return
        self._data = {}


class LampCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'LampCompartment':
        c = LampCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class Lamp:
    def __init__(self):
        self._state_stack = []
        self._context_stack = []
        self.color = "white"
        self.switch_closed = False
        self.__compartment = LampCompartment("Off")
        self.__next_compartment = None
        __frame_event = LampFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = LampFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = LampFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = LampFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def turnOn(self):
        __e = LampFrameEvent("turnOn", None)
        __ctx = LampFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def turnOff(self):
        __e = LampFrameEvent("turnOff", None)
        __ctx = LampFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def getColor(self) -> str:
        __e = LampFrameEvent("getColor", None)
        __ctx = LampFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        return self._context_stack.pop()._return

    def setColor(self, color: str):
        __e = LampFrameEvent("setColor", {"color": color})
        __ctx = LampFrameContext(__e, None)
        self._context_stack.append(__ctx)
        self.__kernel(__e)
        self._context_stack.pop()

    def _state_On(self, __e):
        if __e._message == "<$":
            self.openSwitch()
        elif __e._message == "$>":
            self.closeSwitch()
        elif __e._message == "getColor":
            self._context_stack[-1]._return = self.color
            return
        elif __e._message == "setColor":
            color = __e._parameters["color"]
            self.color = color
        elif __e._message == "turnOff":
            __compartment = LampCompartment("Off", parent_compartment=self.__compartment.copy())
            self.__transition(__compartment)

    def _state_Off(self, __e):
        if __e._message == "getColor":
            self._context_stack[-1]._return = self.color
            return
        elif __e._message == "setColor":
            color = __e._parameters["color"]
            self.color = color
        elif __e._message == "turnOn":
            __compartment = LampCompartment("On", parent_compartment=self.__compartment.copy())
            self.__transition(__compartment)

    def closeSwitch(self):
        self.switch_closed = True

    def openSwitch(self):
        self.switch_closed = False


def main():
    print("=== Test 31: Doc Lamp Basic ===")
    lamp = Lamp()

    # Initially off
    assert lamp.switch_closed == False, "Switch should be open initially"

    # Turn on - should close switch
    lamp.turnOn()
    assert lamp.switch_closed == True, "Switch should be closed after turnOn"

    # Check color
    assert lamp.getColor() == "white", f"Expected 'white', got '{lamp.getColor()}'"

    # Set color
    lamp.setColor("blue")
    assert lamp.getColor() == "blue", f"Expected 'blue', got '{lamp.getColor()}'"

    # Turn off - should open switch
    lamp.turnOff()
    assert lamp.switch_closed == False, "Switch should be open after turnOff"

    print("PASS: Basic lamp works correctly")

if __name__ == '__main__':
    main()

File: output/javascript/tests/31_doc_lamp_basic.js

// Documentation Example: Basic Lamp with enter/exit events

export class LampFrameEvent {
    _message;
    _parameters;

    constructor(message, parameters) {
        this._message = message;
        this._parameters = parameters;
    }
}


export class LampFrameContext {
    event;
    _return;
    _data;

    constructor(event, default_return) {
        this.event = event;
        this._return = default_return;
        this._data = {  };
    }
}


export class LampCompartment {
    state;
    state_args;
    state_vars;
    enter_args;
    exit_args;
    forward_event;
    parent_compartment;

    constructor(state, parent_compartment = null) {
        this.state = state;
        this.state_args = {  };
        this.state_vars = {  };
        this.enter_args = {  };
        this.exit_args = {  };
        this.forward_event = null;
        this.parent_compartment = parent_compartment;
    }

    copy() {
        const c = new LampCompartment(this.state, this.parent_compartment);
        c.state_args = {...this.state_args};
        c.state_vars = {...this.state_vars};
        c.enter_args = {...this.enter_args};
        c.exit_args = {...this.exit_args};
        c.forward_event = this.forward_event;
        return c;
    }
}


export class Lamp {
    _state_stack;
    __compartment;
    __next_compartment;
    _context_stack;
    color = "white";
    switch_closed = false;

    constructor() {
        this._state_stack = [];
        this._context_stack = [];
        this.__compartment = new LampCompartment("Off");
        this.__next_compartment = null;
        const __frame_event = new LampFrameEvent("$>", null);
        this.__kernel(__frame_event);
    }

    __kernel(__e) {
        // Route event to current state
        this.__router(__e);
        // Process any pending transition
        while (this.__next_compartment !== null) {
            const next_compartment = this.__next_compartment;
            this.__next_compartment = null;
            // Exit current state
            const exit_event = new LampFrameEvent("<$", this.__compartment.exit_args);
            this.__router(exit_event);
            // Switch to new compartment
            this.__compartment = next_compartment;
            // Enter new state (or forward event)
            if (next_compartment.forward_event === null) {
                const enter_event = new LampFrameEvent("$>", this.__compartment.enter_args);
                this.__router(enter_event);
            } else {
                // Forward event to new state
                const forward_event = next_compartment.forward_event;
                next_compartment.forward_event = null;
                if (forward_event._message === "$>") {
                    // Forwarding enter event - just send it
                    this.__router(forward_event);
                } else {
                    // Forwarding other event - send $> first, then forward
                    const enter_event = new LampFrameEvent("$>", this.__compartment.enter_args);
                    this.__router(enter_event);
                    this.__router(forward_event);
                }
            }
        }
    }

    __router(__e) {
        const state_name = this.__compartment.state;
        const handler_name = `_state_${state_name}`;
        const handler = this[handler_name];
        if (handler) {
            handler.call(this, __e);
        }
    }

    __transition(next_compartment) {
        this.__next_compartment = next_compartment;
    }

    turnOn() {
        const __e = new LampFrameEvent("turnOn", null);
        const __ctx = new LampFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    turnOff() {
        const __e = new LampFrameEvent("turnOff", null);
        const __ctx = new LampFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    getColor() {
        const __e = new LampFrameEvent("getColor", null);
        const __ctx = new LampFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    setColor(color) {
        const __e = new LampFrameEvent("setColor", {"color": color});
        const __ctx = new LampFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        this._context_stack.pop();
    }

    isSwitchClosed() {
        const __e = new LampFrameEvent("isSwitchClosed", null);
        const __ctx = new LampFrameContext(__e, null);
        this._context_stack.push(__ctx);
        this.__kernel(__e);
        return this._context_stack.pop()._return;
    }

    _state_On(__e) {
        if (__e._message === "<$") {
            this.openSwitch()
        } else if (__e._message === "$>") {
            this.closeSwitch()
        } else if (__e._message === "getColor") {
            this._context_stack[this._context_stack.length - 1]._return = this.color;
            return;
        } else if (__e._message === "isSwitchClosed") {
            this._context_stack[this._context_stack.length - 1]._return = this.switch_closed;
            return;
        } else if (__e._message === "setColor") {
            const color = __e._parameters?.["color"];
            this.color = color
        } else if (__e._message === "turnOff") {
            const __compartment = new LampCompartment("Off", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        }
    }

    _state_Off(__e) {
        if (__e._message === "getColor") {
            this._context_stack[this._context_stack.length - 1]._return = this.color;
            return;
        } else if (__e._message === "isSwitchClosed") {
            this._context_stack[this._context_stack.length - 1]._return = this.switch_closed;
            return;
        } else if (__e._message === "setColor") {
            const color = __e._parameters?.["color"];
            this.color = color
        } else if (__e._message === "turnOn") {
            const __compartment = new LampCompartment("On", this.__compartment.copy());
            this.__transition(__compartment);
            return;
        }
    }

    closeSwitch() {
                    this.switch_closed = true
    }

    openSwitch() {
                    this.switch_closed = false
    }
}

function main() {
    console.log("=== Test 31: Doc Lamp Basic ===");
    const lamp = new Lamp();

    // Initially off
    if (lamp.isSwitchClosed() !== false) {
        throw new Error("Switch should be open initially");
    }

    // Turn on - should close switch
    lamp.turnOn();
    if (lamp.isSwitchClosed() !== true) {
        throw new Error("Switch should be closed after turnOn");
    }

    // Check color
    if (lamp.getColor() !== "white") {
        throw new Error(`Expected 'white', got '${lamp.getColor()}'`);
    }

    // Set color
    lamp.setColor("blue");
    if (lamp.getColor() !== "blue") {
        throw new Error(`Expected 'blue', got '${lamp.getColor()}'`);
    }

    // Turn off - should open switch
    lamp.turnOff();
    if (lamp.isSwitchClosed() !== false) {
        throw new Error("Switch should be open after turnOff");
    }

    console.log("PASS: Basic lamp works correctly");
}

main();
